state/upgrades.go

```diff
@@ -65,185 +41,124 @@ func runForAllEnvStates(st *State, runner func(st *State) error) error {
-// it should be Attaching; otherwise it is Attached.
-func upgradingFilesystemStatus(st *State, filesystem Filesystem) (status.Status, error) {
-	if _, err := filesystem.Info(); errors.IsNotProvisioned(err) {
-		return status.Pending, nil
+	more, err = stripLocalFromFields(st, usermodelnameC, "_id")
+	if err != nil {
+		return err
 	}
-	attachments, err := st.FilesystemAttachments(filesystem.FilesystemTag())
+	ops = append(ops, more...)
+
+	more, err = stripLocalFromFields(st, controllerUsersC, "_id", "user", "createdby")
 	if err != nil {
-		return "", errors.Trace(err)
+		return err
 	}
-	for _, attachment := range attachments {
-		_, err := attachment.Info()
-		if errors.IsNotProvisioned(err) {
-			return status.Attaching, nil
-		}
+	ops = append(ops, more...)
+
+	more, err = stripLocalFromFields(st, modelUsersC, "_id", "user", "createdby")
+	if err != nil {
+		return err
+	}
+	ops = append(ops, more...)
+
+	more, err = stripLocalFromFields(st, permissionsC, "_id", "subject-global-key")
+	if err != nil {
+		return err
 	}
-	return status.Attached, nil
+	ops = append(ops, more...)
+
+	more, err = stripLocalFromFields(st, modelUserLastConnectionC, "_id", "user")
+	if err != nil {
+		return err
+	}
+	ops = append(ops, more...)
+	return st.runRawTransaction(ops)
 }
 
-// MigrateSettingsSchema migrates the schema of the settings collection,
-// moving non-reserved keys at the top-level into a subdoc, and introducing
-// a top-level "version" field with the initial value matching txn-revno.
-//
-// This migration takes place both before and after model-uuid migration,
-// to get the correct txn-revno value.
-func MigrateSettingsSchema(st *State) error {
-	coll, closer := st.getRawCollection(settingsC)
+func stripLocalFromFields(st *State, collName string, fields ...string) ([]txn.Op, error) {
+	coll, closer := st.getRawCollection(collName)
 	defer closer()
+	upgradesLogger.Infof("migrating document fields of the %s collection", collName)
 
-	upgradesLogger.Debugf("migrating schema of the %s collection", settingsC)
 	iter := coll.Find(nil).Iter()
 	defer iter.Close()
-
 	var ops []txn.Op
 	var doc bson.M
 	for iter.Next(&doc) {
-		if !settingsDocNeedsMigration(doc) {
-			continue
-		}
-
+		// Get a copy of the current doc id so we can see if it has changed.
 		id := doc["_id"]
-		txnRevno := doc["txn-revno"].(int64)
-
-		// Remove reserved attributes; we'll move the remaining
-		// ones to the "settings" subdoc.
-		delete(doc, "model-uuid")
-		delete(doc, "_id")
-		delete(doc, "txn-revno")
-		delete(doc, "txn-queue")
-
-		// If there exists a setting by the name "settings",
-		// we must remove it first, or it will collide with
-		// the dotted-notation $sets.
-		if _, ok := doc["settings"]; ok {
-			ops = append(ops, txn.Op{
-				C:      settingsC,
-				Id:     id,
-				Assert: txn.DocExists,
-				Update: bson.D{{"$unset", bson.D{{"settings", 1}}}},
-			})
+		newId := id
+
+		// Take a copy of the current doc fields.
+		newDoc := make(bson.M)
```

Comment:
Using bson.M instead of bson.D will rearrange field ordering which makes me nervous. See https://bugs.launchpad.net/juju-core/+bug/1451674

We had helpers in this file in 1.25 do make reading and updating bson.D based documents easier. See: https://github.com/juju/juju/blame/1.25/state/upgrades.go#L1188